sound/soc/sof/intel/hda-dai.c

```diff
@@ -184,7 +180,6 @@ static int hda_link_dma_params(struct hdac_ext_stream *hext_stream,
 	struct hdac_ext_link *link;
 	unsigned int format_val;
 
-	snd_hdac_ext_stream_decouple(bus, hext_stream, true);
```

Comment:
@ujfalusi this code was loosely inspired from the code in sound/hda/ext/hdac_ext_stream.c

That starting point was already rather questionable and it's not clear what the adaptation kept.

This is a rather dangerous can of worms you're opening.

Reply:
@ujfalusi the other problem is during suspend, we'd call snd_hdac_ext_stream_release() which would end up coupling them back but we'd never call stream_assign after resume

Reply:
@plbossart, @ranj063, right, hrm, interesting and confusing...
We are using 'open coded' `snd_hdac_ext_stream_assign()` implemented as `hda_link_stream_assign()` and we use the `snd_hdac_ext_stream_release()` to release it on STOP/PAUSE. But where we release it we are also dropping the dma_data for the cpu_dai, so the next `hda_link_dma_hw_params()` will again going to fetch a new hext_stream, so in case of suspend we release (which _might_ coupling back, but it might not if `hext_stream->hstream.opened == true`, which might be as long as we have the stream open, I think).

The `hext_stream->decoupled` is only changed by `snd_hdac_ext_stream_decouple_locked()` but it is unconditional internally regarding to the saved decoupled flag, so it will go ahead and check the hardware itself if there is a need for the change. If the hext_stream is already decoupled then we can save register reads, the two if case and the assignment by forward checking the flag.

But I still don't see why we would need to do the decouple twice since the second one is only going to read the PPCTL, do the two if and set the hext_stream->decoupled from `true` to `true`

Reply:
On the other hand: I believe the reason why we had the unconditional call to `snd_hdac_ext_stream_decouple()` in `hda_link_dma_hw_params()` was to force the change when we came out from suspend for example since in our assign we would have skipped calling the `snd_hdac_ext_stream_decouple_locked()` but the hardware actually switched back to reset state, which is coupled.

I do think that this change is sound.

Reply:
@ujfalusi sorry  not following. But havent you removed the decoupling in hw_params now? So we wouldnt actually decouple once w resume right?

Reply:
@ujfalusi right now it makes sense. We recently changed this logic to set the dma_data to NULL during stop always. I got misled by the comment https://github.com/thesofproject/linux/blob/bfe718118ab3db38df3a69585ee74d461ca43f49/sound/soc/sof/intel/hda-dai.c#L224.

Maybe you should remove this comment and the check in this PR too

Reply:
I don't get what you want to remove @ranj063 

 hda_link_dma_hw_params() is called in the prepare() step, which can be called either
a) in case of a xrun. In this case we don't want to re-allocate the stream
b) in case of a system suspend, in which case we do want to re-allocate.

Reply:
It's probably only the comment that needs to be reworded as "don't allocate the stream in case of xrun"

Reply:
@plbossart, @ranj063, or just dropped? It would be reduced to state the obvious.